### PR TITLE
KAFKA-16557 Implemented OffsetFetchRequestState toStringBase and added a test for it in CommitRequestManagerTest

### PR DESCRIPTION
**Pull request by brenden20**
I changed OffsetFetchRequestState toString to toStringBase. I also added the instance fields from OffsetFetchRequestState and RetriableState to the toStringBase method. Within RetriableState, I also added a getter method for expirationTimeMs for testing purposes.

I added a new test to CommitRequestManagerTest that tests the new toStringBase method. In the test, a string variable 'target' is instantiated and will be compared to the new toStringBase method. The test calls OffsetFetchRequestState toStringBase() and compares it to the target string. All tests in CommitRequestManagerTest pass.

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 1]
```diff
@@ -227,8 +275,8 @@ public void testPollEnsureEmptyPendingRequestAfterPoll() {
         CommitRequestManager commitRequestManager = create(true, 100);
         when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(mockedNode));
         Map<TopicPartition, OffsetAndMetadata> offsets = Collections.singletonMap(
-            new TopicPartition("topic", 1),
-            new OffsetAndMetadata(0));
+                new TopicPartition("topic", 1),
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: could you undo these changes that aren't related to the pr?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 2]
```diff
@@ -880,6 +880,10 @@ boolean retryTimeoutExpired(long currentTimeMs) {
         }
 
         abstract void removeRequest();
+
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:883]: add a comment // Visible for testing

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 3]
```diff
@@ -1090,14 +1094,15 @@ private void chainFuture(
...
+            return super.toStringBase() +
+                    ", memberInfo=" + memberInfo +
+                    ", expirationTimeMs=" + (expirationTimeMs().isPresent() ? expirationTimeMs() : "undefined") +
+                    ", isExpired=" + isExpired +
+                    ", requestedPartitions=" + requestedPartitions +
                     ", future=" + future +
-                    ", " + toStringBase() +
-                    '}';
+                    ", memberId=" + memberInfo.memberId.orElse("undefined") +
+                    ", memberEpoch=" + (memberInfo.memberEpoch.isPresent() ? memberInfo.memberEpoch : "undefined");
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: My preference would be to add a `toString()` to the `MemberInfo` class. WDYT?

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks for the PR, @brenden20!

I'd never really thought to add a unit test for `toString()`, but it's not a bad idea 👍 

Just to echo @philipnee, revert any unnecessary whitespace/indentation changes.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 2 above)
Comment (brenden20) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:883]: Got that added in a new commit!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java (hunk 3 above)
Comment (brenden20) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java]: Thank you for the suggestion, I added a toString() method to MemberInfo in my latest commit.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java (hunk 1 above)
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: When I try to change it back, it says that the file is the same. I copied the parts of the test that I did not change from a different branch but Git is saying the files are the same still. Not too sure why it showed I made indentations, they are not showing up on my end in Intellij.

**Comment by kirktrue** [comment]
@brenden20, as mentioned on another one of your PRs, there's a checkstyle violation here. You can run this command locally to avoid waiting for the CI infrastructure to catch it:

```
./gradlew check -x test
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java (hunk 1 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: i wonder if intellij's auto indentation automatically changes the indentation when copying the content over.  double check the indentation for these files they should be there.  it could be easier to use a text editor like vim/emac to edit out these unintended changes.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 4]
```diff
@@ -123,6 +126,47 @@ public void setup() {
         this.props.put(VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
     }
 
+    @Test
+    public void testOffsetFetchRequestStateToStringBase() {
+        final long retryBackoffMs = 10;
+        final long retryBackoffMaxMs = 100;
+        final long expirationTimeMs = 1000;
+        ConsumerConfig config = mock(ConsumerConfig.class);
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:129]: try be consistent with the syntax.  either use final for the config and memberinfo or just not use final for the longs.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 5]
```diff
@@ -123,6 +126,47 @@ public void setup() {
...
+
+        this.requestState = new RequestState(
+                logContext,
+                "CommitRequestManager",
+                retryBackoffMs,
+                retryBackoffMaxMs);
+
+        String target = requestState.toStringBase() +
+                ", memberInfo={" + offsetFetchRequestState.memberInfo +
+                "}, expirationTimeMs=" + (offsetFetchRequestState.expirationTimeMs().isPresent() ? offsetFetchRequestState.expirationTimeMs() : "undefined") +
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:162]: i think we are expecting an expirationTimeMs so just use `expirationTimeMs=" + expirationTimeMs`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 6]
```diff
@@ -123,6 +126,47 @@ public void setup() {
...
+        );
+
+        this.requestState = new RequestState(
+                logContext,
+                "CommitRequestManager",
+                retryBackoffMs,
+                retryBackoffMaxMs);
+
+        String target = requestState.toStringBase() +
+                ", memberInfo={" + offsetFetchRequestState.memberInfo +
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: we can just test the values against L131-135, i think target shoud just be:
memberInfo={memberInfo}, as well as other fields below.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 7]
```diff
@@ -123,6 +126,47 @@ public void setup() {
...
+                config,
+                coordinatorRequestManager,
+                offsetCommitCallbackInvoker,
+                "groupId",
+                Optional.of("groupInstanceId"),
+                metrics);
+
+        this.offsetFetchRequestState = commitRequestManager.new OffsetFetchRequestState(
+                mock(Set.class),
+                retryBackoffMs, retryBackoffMaxMs, expirationTimeMs,
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: let break this into separated line just to be consistent with L154-158.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java [hunk 8]
```diff
@@ -1289,5 +1293,11 @@ static class MemberInfo {
             this.memberId = Optional.empty();
             this.memberEpoch = Optional.empty();
         }
+
+        @Override
+        public String toString() {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:1298]: Add the class name.  reformat it to
```
MemberInfo(memberId=..., memberEpoch=...)
```

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 9]
```diff
@@ -106,6 +106,9 @@ public class CommitRequestManagerTest {
     private OffsetCommitCallbackInvoker offsetCommitCallbackInvoker;
     private final Metrics metrics = new Metrics();
     private Properties props;
+    private CommitRequestManager.OffsetFetchRequestState offsetFetchRequestState;
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: the `commitRequestManager`, `offsetFetchRequestState` and `requestState` can all stay in the scope of `testOffsetFetchRequestStateToStringBase` no?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 10]
```diff
@@ -123,6 +126,47 @@ public void setup() {
...
+                Optional.of("groupInstanceId"),
+                metrics);
+
+        this.offsetFetchRequestState = commitRequestManager.new OffsetFetchRequestState(
+                mock(Set.class),
+                retryBackoffMs, retryBackoffMaxMs, expirationTimeMs,
+                memberInfo
+        );
+
+        this.requestState = new RequestState(
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java]: this should be `new RetriableRequestState` since offsetFetchRequestState implements the retriable one?

**Comment by philipnee** [comment]
hi @brenden20 - left some comments. lmk what do you think.

**Comment by brenden20** [comment]
@philipnee I have implemented all suggested changes except for the whitespace issue, I am going to use Emacs to try to figure it out

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 11]
```diff
@@ -123,6 +123,46 @@ public void setup() {
...
+
+        CommitRequestManager.OffsetFetchRequestState offsetFetchRequestState = commitRequestManager.new OffsetFetchRequestState(
+                new HashSet<>(),
+                retryBackoffMs,
+                retryBackoffMaxMs,
+                1000,
+                memberInfo
+        );
+
+        RequestState requestState = new RequestState(
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:154]: we should use `RetriableRequestState` right? such that we can get the expirationTimeMs

Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:154]: RetriableRequestState is abstract, I do not think that I can instantiate that class directly

Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:154]: you are right, scratch that.

Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:154]: Gotcha, I will mark this suggestion as resolved then.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 12]
```diff
@@ -123,6 +123,46 @@ public void setup() {
...
+                retryBackoffMs,
+                retryBackoffMaxMs);
+
+        String target = requestState.toStringBase() +
+                ", memberInfo=" + offsetFetchRequestState.memberInfo +
+                ", expirationTimeMs=" + (offsetFetchRequestState.expirationTimeMs().isPresent() ? offsetFetchRequestState.expirationTimeMs() : "undefined") +
+                ", isExpired=" + offsetFetchRequestState.isExpired +
+                ", requestedPartitions=" + offsetFetchRequestState.requestedPartitions +
+                ", future=" + offsetFetchRequestState.future();
+
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:166]: can we also assert no exception can be thrown for calling toString() - just to ensure no lousy toString().

**Review by kirktrue — APPROVED** [review]
LGTM! Can we move this from Draft to Ready to review?

**Comment by brenden20** [comment]
Marked ready to review @kirktrue

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java [hunk 13]
```diff
@@ -123,6 +124,50 @@ public void setup() {
...
+        RequestState requestState = new RequestState(
+                logContext,
+                "CommitRequestManager",
+                retryBackoffMs,
+                retryBackoffMaxMs);
+
+        String target = requestState.toStringBase() +
+                ", memberInfo=" + memberInfo +
+                ", expirationTimeMs=" + (offsetFetchRequestState.expirationTimeMs().isPresent() ? offsetFetchRequestState.expirationTimeMs() : "undefined") +
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:162]: we can directly do `expirationTimeMs=1000` here. in this case we do that we don't need to change the scope of `expirationTimeMs()`

Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/CommitRequestManagerTest.java:162]: Looking at the code, I am able to do this, however it will cause an issue with OffsetFetchRequestState.toStringBase(). ```expirationTimeMs()``` is needed to implement the ```toStringBase()``` method, since ```OffsetFetchRequestState``` cannot access ```RetriableRequestState.expirationTimeMs``` since it is a private field.

**Comment by brenden20** [comment]
Had a lot of merge issues, was taking too much time to fix so I made a new branch with my changes. New PR is here: https://github.com/apache/kafka/pull/16291
